core/classifiers/CodeClassifier/CodeClassifier.py

```diff
@@ -0,0 +1,695 @@
+                        'tokens': list(str). Token list of program.
+                    } for each program in training_data.
+                }
+        token_to_id: dict. A dictionary which maps each token to a unique ID.
+        K: int. Model parameter 'K' of winnowing.
+
+    Returns:
+        dict. A dictionary containing training data with additional key
+        'k_gram_hashes' for each data point, storing k-gram hashes corresponding
+        to that data point. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                } for each program in training_data.
+            }
+    """
+    for program_id in training_data:
+        training_data[program_id]['k_gram_hashes'] = (
+            winnowing.k_gram_hash_generator(
+                training_data[program_id]['tokens'], token_to_id, K))
+    return training_data
+
+
+def generate_program_fingerprints(training_data, T, K):
+    """Generate document fingerprints for all programs in entire dataset.
+
+    Args:
+        training_data: dict. A dictionary containing training data. Structure of
+            training_data is as follows,
+                {
+                    ID: {
+                        'source': str. Source code of program.
+                        'class': int. Associated feedback class with program.
+                        'tokens': list(str). Token list of program.
+                        'k_gram_hashes': list(int). K-gram hash values of
+                            program.
+                    } for each program in training_data.
+                }
+        T: int. Model parameter 'T' of winnowing.
+        K: int. Model parameter 'K' of winnowing.
+
+    Returns:
+        dict. A dictionary containing training data with additional key
+        'fingerprint' for each program which stores generated fingerprint
+        of program. Structure of dict is as follows,
+            {
+                ID: {
+                    'source': str. Source code of program.
+                    'class': int. Associated feedback class with program.
+                    'tokens': list(str). Token list of program.
+                    'k_gram_hashes': list(int). K-gram hash values of program.
+                    'fingerprint': list(int). Extracted fingerprint of program.
+                } for each program in training_data.
+            }
+    """
+    window_size = T - K + 1
+    for program_id in training_data:
+        training_data[program_id]['fingerprint'] = (
+            winnowing.get_fingerprint_from_hashes(
+                training_data[program_id]['k_gram_hashes'], window_size))
+    return training_data
+
+
+def calc_jaccard_index(set_a, set_b):
```

Comment:
The scenario you mentioned is desirable behavior. But that is still a set since first 3 and second 3 do not originate from same k-gram hash. What I mean is that first 3 may originate from substring X and next 3 also originates from X but that means that program has substring X twice. So you can't say remove second 3 because it is distinct from first 3 in location from where it originates in program. So they are still sets, just don't think of them as numbers. That is why they need to be considered distinct. So it matches with those program which have X substirng twice (Otherwise it would match with those program which have substring X once as well as twice and that is not desirable). In this context jaccard is still acceptable.

These type of set operations are permissible in [SQL](https://www.mssqltips.com/sqlservertip/1387/union-vs-union-all-in-sql-server/). I am doing intersect-all and union-all. here.